MudAutocomplete: Scroll to selected item when not strict

digitaldirk opened the pull request:
Now when `strict` is `false` the selected item will be scrolled to when opening the Autocomplete. 

## Description
When selecting an item with strict set to false, then clicking the Autocomplete the list would just open to the first item.  This behavior is the opposite of what MudSelect does. This PR matches the behavior and will scroll to selected item.

_Do we want to add a bool parameter that turns the auto scroll on/off?_

(You can compare the current behavior: select an item, then reopen the select/autocomplete)
https://www.mudblazor.com/components/select#visual-playground
https://www.mudblazor.com/components/autocomplete#strict-mode

## How Has This Been Tested?
Visually tested with current tests passing.

## Type of Changes
Could be a breaking change if user implemented custom scroll logic?

**Before:**
![beforescroll](https://github.com/user-attachments/assets/c930822d-826e-43dc-bf3f-f29e8e4eb74c)
**After:**
![afterscroll](https://github.com/user-attachments/assets/9f0c5af1-9d7d-44b5-a004-052a2ceb2c8d)

## Checklist
- [X] The PR is submitted to the correct branch (`dev`).
- [X] My code follows the code style of this project.
- [ ] I've added relevant tests.

digitaldirk committed:
Scroll to selected item when not strict

Match behavior to MudSelect

Anu6is commented:
I'm not sure that this would be enough. Remember that the AutoComplete, by default, shows only the first 10 items. So, if the user selects the 12th item in the list by typing, when they re-open the list the selected item won't be rendered in the initial 10 items meaning it can't be scrolled to.

digitaldirk committed:
Update MudAutocomplete.razor.cs

Add logic to account for MaxItems being not null

digitaldirk commented:
With latest commit this now accounts for non-null `MaxItems`. The logic gets a range of `MaxItems` based on the index of the seleected value.

This example is the using the default of 10. Washington has 3 states under it so in this case, 6 states are above it (3 + selected + 6 = 10). But when in the middle of the list, there would be 5 states below the selected state, and 4 above (for a total of 10 states).
![MaxItemsFix](https://github.com/user-attachments/assets/b77db04c-1552-4088-82c1-fb5e2bf3f904)

digitaldirk committed:
Fix index bug

Account for MaxItems > total # of items

digitaldirk commented:
I will fix the current failing tests and add test(s) for the new functionality but first I would appreciate feedback before I do that.

Do we want to be able to toggle on/off this functionality? I can see some not wanting this, but I think it should be on be default because that is the way MudSelect works.

digitaldirk commented:
Sorry for the pings but I'd like feedback before continuing to work on this.

@henon @danielchalmers @versile2 

Thank you for your time :)

versile2 commented:
I can't speak for others but I like auto scrolling to it and I don't see a need to add yet another option. There's no harm in it and I can't think of a scenario where scrolling automatically to the selected item would cause an issue. I do know there is a pending PR to fix some functionality so I would wait until that is merged then pull it into this one before continuing. Ultimately the decision falls to others but that is my opinion.

V

henon commented:
I am for adding this too. I would even go so far as to call the current behavior a bug.

danielchalmers commented:
Good thinking about compatibility concerns, but I agree this is a bug and can be fixed without any extra steps

danielchalmers commented:
Good thinking about compatibility concerns, but I agree this is a bug and can be fixed without any extra steps

danielchalmers commented:
Good thinking about compatibility concerns, but I agree this is a bug and can be fixed without any extra steps

digitaldirk committed:
Merge branch 'MudBlazor:dev' into MudAutocomplete-ScrollToSelectedNotStrict

digitaldirk committed:
Fix AutocompleteStrictFalseTest for new scroll to logic

New scroll to logic now shows different items/different item indexes depending on what is searched for. Changed tests to now correctly find indexes/expected items.

digitaldirk committed:
Fix SonarQube Major Code Smell 

Combine if statement

digitaldirk commented:
Thank you all for the feedback @Anu6is @versile2 @henon @danielchalmers 

**Fixed the failing tests and the SonarQube Major issue, but could use some help on:**
1. https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&pullRequest=11049&id=MudBlazor_MudBlazor&open=AZWq4zV_GkTDPuUTQLzl&tab=code This one is going over my head - could I get an explanation or help on fixing this one? Why is it complaining about using Value != null - we use that on https://github.com/digitaldirk/MudBlazor/blob/a427d59ddf6b47bd5bd6a288374fbf16d7ce88db/src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs#L703 as well.
2. Sorry about the formatting and many commits (on a work network with firewalls so I have to edit the commits raw in the browser 😓) Can someone format my changes?

All done besides any action on the above 2 points

versile2 commented:
> Thank you all for the feedback @Anu6is @versile2 @henon @danielchalmers
> 
> **Fixed the failing tests and the SonarQube Major issue, but could use some help on:**
> 
>     1. https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&pullRequest=11049&id=MudBlazor_MudBlazor&open=AZWq4zV_GkTDPuUTQLzl&tab=code This one is going over my head - could I get an explanation or help on fixing this one? Why is it complaining about using Value != null - we use that on https://github.com/digitaldirk/MudBlazor/blob/a427d59ddf6b47bd5bd6a288374fbf16d7ce88db/src/MudBlazor/Components/Autocomplete/MudAutocomplete.razor.cs#L703 as well.
> 
>     2. Sorry about the formatting and many commits (on a work network with firewalls so I have to edit the commits raw in the browser 😓) Can someone format my changes?
> 
> 
> All done besides any action on the above 2 points

1) because it's a generic type, generally just putting default<T> does the trick. I'm not good at explaining though so if someone else wants to please do.
2) You will have to fix alignment and spacing issues. Do you not have access at home to fix? Only core members can "edit" your PR and they have way too much to do. As for commits make as many as you like we only look at the end result.

digitaldirk committed:
Merge branch 'MudBlazor:dev' into MudAutocomplete-ScrollToSelectedNotStrict

versile2 commented:
Are you able to update the spacing issues and what not?

versile2 commented:
> > Are you able to update the spacing issues and what not?
> 
> Yes, will be able to do the last steps this week.

Awesome, I think this is a value bug fix, hit me up on Discord dm if you have any questions.

V

digitaldirk committed:
Merge branch 'MudBlazor:dev' into MudAutocomplete-ScrollToSelectedNotStrict

digitaldirk committed:
Fix sonar cloud issue

digitaldirk committed:
Merge branch 'MudBlazor:dev' into MudAutocomplete-ScrollToSelectedNotStrict

digitaldirk committed:
Merge branch 'dev' into MudAutocomplete-ScrollToSelectedNotStrict

digitaldirk committed:
fix sonar cloud issue

digitaldirk committed:
Update AutocompleteTests.cs to fix whitespace

Fix whitespace errors

digitaldirk committed:
Update MudAutocomplete.razor.cs fix whitespace

fix whitespace errors

digitaldirk committed:
Update AutocompleteTests.cs fix whitespace

fix whitespace

digitaldirk commented:
Sorry for the downtime on this one!

danielchalmers commented:
https://github.com/user-attachments/assets/edb9fe6a-c4d4-454b-bf7a-64eff6cb5ef4

Is the list supposed to be limited based on the selection? You can only reach elements that are around the selected one

digitaldirk commented:
> 
> https://github.com/user-attachments/assets/edb9fe6a-c4d4-454b-bf7a-64eff6cb5ef4
> 
> Is the list supposed to be limited based on the selection? You can only reach elements that are around the selected one

Yes it is limited by max items, I chose to split the results half above/below.

versile2 reviewed (approved):
Logic change is minimal. I agree with splitting the difference. Test updated to include MaxItems change when string is false.
